src/backend/columnar/columnar_metadata.c

```diff
@@ -218,6 +219,117 @@ InitColumnarOptions(Oid regclass)
 }
 
 
+/*
+ * ExtractColumnarOptions - split inoutOptions into two lists, one for
+ * columnar options and another for everything else.
+ */
+List *
+ExtractColumnarRelOptions(List *inOptions, List **outColumnarOptions)
+{
+	List *columnarOptions = NIL;
+	List *otherOptions = NIL;
+	ListCell *lc;
+
+	foreach (lc, inOptions)
+	{
+		DefElem *elem = castNode(DefElem, lfirst(lc));
+
+		if (elem->defnamespace != NULL &&
+			strcmp(elem->defnamespace, COLUMNAR_RELOPTION_NAMESPACE) == 0)
+		{
+			columnarOptions = lappend(columnarOptions, elem);
+		}
+		else
+		{
+			otherOptions = lappend(otherOptions, elem);
+		}
+	}
+
+	*outColumnarOptions = columnarOptions;
+
+	return otherOptions;
+}
+
+
+/*
+ * SetColumnarRelOptions - apply the list of DefElem options to the relation.
+ */
+void
+SetColumnarRelOptions(RangeVar *rv, List *reloptions)
+{
+	ColumnarOptions options;
+
+	if (reloptions == NIL)
+	{
+		return;
+	}
+
+	Relation rel = relation_openrv(rv, AccessShareLock);
+	Oid relid = RelationGetRelid(rel);
+	relation_close(rel, NoLock);
+
+	/* get existing or default options */
+	if (!ReadColumnarOptions(relid, &options))
+	{
+		/* if extension doesn't exist, just return */
+		return;
+	}
+
+	ListCell *lc;
+	foreach (lc, reloptions)
+	{
+		DefElem *elem = castNode(DefElem, lfirst(lc));
+
+		if (elem->defnamespace == NULL ||
+			strcmp(elem->defnamespace, COLUMNAR_RELOPTION_NAMESPACE) != 0)
+		{
+			ereport(ERROR, (errmsg("columnar options must have the prefix \"%s\"",
+								   COLUMNAR_RELOPTION_NAMESPACE)));
+		}
+
+		if (strcmp(elem->defname, "chunk_group_row_limit") == 0)
+		{
+			options.chunkRowCount = (elem->arg == NULL) ?
+				columnar_chunk_group_row_limit : defGetInt64(elem);
+		}
+		else if (strcmp(elem->defname, "stripe_row_limit") == 0)
+		{
+			options.stripeRowCount = (elem->arg == NULL) ?
+				columnar_stripe_row_limit : defGetInt64(elem);
+		}
+		else if (strcmp(elem->defname, "compression") == 0)
+		{
+			if (elem->arg == NULL)
+			{
+				options.compressionType = columnar_compression;
+			}
+			else
+			{
+				char *compressionStr = defGetString(elem);
+				CompressionType compressionType = ParseCompressionType(compressionStr);
+
+				if (options.compressionType == COMPRESSION_TYPE_INVALID)
```

Comment:
when setting the compression type to an unknown/invalid compression type, currently the backend crashes. eg.
```sql
ALTER TABLE c SET (columnar.compression = fooz);
```
Causes a crash when `c` is using the columnar access method.

I think it is due to the comparison we are doing here.
We will either need to set `options.compressionType` directly above, or we will want to compare against `compressionType` here.

This also indicates we don't have a test for setting the option to something bogus. Might be a good moment to add one.

Reply:
Fixed. We do have one test for that, which caught the problem. Thank you.